orc/src/main/java/org/apache/iceberg/orc/OrcIterable.java

```diff
@@ -130,4 +140,25 @@ public T next() {
     }
   }
 
+  private static class OrcBatchIterator<T> implements Iterator<T> {
```

Comment:
This needs to be a `CloseableIterator` to avoid leaking an open file. I'm also wondering if we could replace this implementation with a method like `CloseableIterable.transform` for `CloseableIterator`. All this is doing is calling a method on the result of another `Iterator`.

Reply:
This `Iterator` is converted to a `CloseableIterator` at https://github.com/apache/iceberg/blob/6a5e8c53d3950922b8f6bd753329b7d943ac423d/orc/src/main/java/org/apache/iceberg/orc/OrcIterable.java#L89
I too want to avoid creating the `OrcBatchIterator` class, however I am unable to make `Iterators.transform` work here because of the capture type of `batchReaderFunction`.

![image](https://user-images.githubusercontent.com/6961317/87212841-92816e80-c2d5-11ea-9fb8-77f16f12206f.png)

I am also not able to cast the `batchReaderFunction` to the generic type used by `OrcIterable`.

Reply:
What about `Iterators.transform(rowBatchIterator, batchReaderFunction.apply(readOrcSchema)::read)`? That should allow the compiler to convert it to a Function.

Reply:
> This Iterator is converted to a CloseableIterator at

Yes, it is converted to a `CloseableIterator`, but that iterator's close method won't do anything unless this is `Closeable`. Here's where it delegates:

```java
      @Override
      public void close() throws IOException {
        if (iterator instanceof Closeable) {
          ((Closeable) iterator).close();
        }
      }
```

So if this `Iterator` doesn't implement `Close` to pass the close to `batchIter`, then `batchIter` won't get closed.

Alternatively, you could add a `transform` factory method to `CloseableIterator`:

```java
  static <S, T> CloseableIterator<T> transform(CloseableIterator<S> iter, Function<S, T> transform) {
    return new CloseableIterator<T>() {
      @Override
      public void close() throws IOException {
        iter.close();
      }

      @Override
      public boolean hasNext() {
        return iter.hasNext();
      }

      @Override
      public T next() {
        return transform.apply(iter.next());
      }
    };
  }
```

Reply:
I remember trying this too. `OrcBatchReader` is a functional interface, so it should not have issues converting the reader to a Function. The issue seems to be the capture type.

![image](https://user-images.githubusercontent.com/6961317/87213393-7e3f7080-c2d9-11ea-9797-d7c717867391.png)

Reply:
Okay, I am able to make it work if I assign the `transformed` iterator to `Iterator` instead of `Iterator<T>`.

Reply:
Fixed in https://github.com/apache/iceberg/pull/1189/commits/7d0cf1cafec9ff42d8e4cb5019185af00ff80d58

Added `CloseableIterator.transform`, updated `VectorizedRowBatchIterator` to use `CloseableIterator`, and also fixed the issue with incorrect use of `CloseableIterator.withClose`.

Reply:
PaymentFlowInfoObject 那邊～